dotnet/src/Extensions/PromptTemplate.Handlebars/Helpers/SystemHelpers/KernelSystemHelpers.cs

```diff
@@ -0,0 +1,169 @@
+            options.CustomHelperPaths = helperOptions.CustomHelperPaths;
+        });
+    }
+
+    /// <summary>
+    /// Register all system helpers.
+    /// </summary>
+    /// <param name="handlebarsInstance">The <see cref="IHandlebars"/>-instance.</param>
+    /// <param name="variables">Dictionary of variables maintained by the Handlebars context.</param>
+    /// <exception cref="KernelException">Exception thrown when a message does not contain a defining role.</exception>
+    private static void RegisterSystemHelpers(
+        IHandlebars handlebarsInstance,
+        KernelArguments variables)
+    {
+        // TODO [@teresaqhoang]: Issue #3947 Isolate Handlebars Kernel System helpers in their own class
+        // Should also consider standardizing the naming conventions for these helpers, i.e., 'Message' instead of 'message'
+        handlebarsInstance.RegisterHelper("message", (writer, options, context, arguments) =>
+        {
+            var parameters = arguments[0] as IDictionary<string, object>;
+
+            // Verify that the message has a role
+            if (!parameters!.ContainsKey("role"))
+            {
+                throw new KernelException("Message must have a role.");
+            }
+
+            writer.Write($"<{parameters["role"]}~>", false);
+            options.Template(writer, context);
+            writer.Write($"</{parameters["role"]}~>", false);
+        });
+
+        handlebarsInstance.RegisterHelper("set", (writer, context, arguments) =>
+        {
+            var name = string.Empty;
+            object value = string.Empty;
+            if (arguments[0].GetType() == typeof(HashParameterDictionary))
+            {
+                // Get the parameters from the template arguments
+                var parameters = arguments[0] as IDictionary<string, object>;
+                name = (string)parameters!["name"];
+                value = parameters!["value"];
+            }
+            else
+            {
+                name = arguments[0].ToString();
+                value = arguments[1];
+            }
+
+            if (variables.TryGetValue(name, out var argVal))
+            {
+                variables[name] = value;
+            }
+            else
+            {
+                variables.Add(name, value);
+            }
+        });
+
+        handlebarsInstance.RegisterHelper("get", (in HelperOptions options, in Context context, in Arguments arguments) =>
+        {
+            if (arguments[0].GetType() == typeof(HashParameterDictionary))
+            {
+                var parameters = arguments[0] as IDictionary<string, object>;
+                return variables[(string)parameters!["name"]];
+            }
+
+            return variables[arguments[0].ToString()];
+        });
+
+        handlebarsInstance.RegisterHelper("json", (in HelperOptions options, in Context context, in Arguments arguments) =>
+        {
+            object objectToSerialize = arguments[0];
```

Comment:
In some of these helpers, the value from the arguments is checked for null, which suggests they can be null. In other places, though, like this one, we're just dereferencing.

Reply:
Added null and empty argument checks for the "json" and "equals" helpers